include/EntryPointConfirmOptIn.php

```diff
@@ -43,40 +43,104 @@
 /**
  * Class confirm_opt_in
  */
-class ConfirmOptIn
+class EntryPointConfirmOptIn
```

Comment:
So this class is a view. The changes in this class is starting to look like a controller. I created ConfirmedOptIn to differentiate the display logic from the entry point logic. Perhaps I should have named it ConfirmedOptInView.

These changes will prevent someone from customising the process in the custom folder. The entry point is the file which could be overridden here. Which would let a third party customise how the opt in process works.

Also, if the email address logic is in the basic template, why do we need to move the confirmed opt in to a locate outside the basic template?

Reply:
So this class is a view. - No, not anymore. This class handle the entry points, (see constructor). So it's a controller.

Reply:
"No, not anymore" is not an adequate answer.

It is best practice to separate our business logic from the display logic. We should have a controller and a view class.

I am still not sure why we are moving the logic away from the basic sugar object. Can you please clarify? What was the issue here?

Reply:
I don't think there is any reason to add it to a path like 'sugarObject/template'. Even if it is not a sugar object nor template. You added a new PHP class into a template folder which was not a SugarObject. the PHP classes are PHP classes but NOT templates. also it was not a sugar objects.

Reply:
When I initially added this code, I wasn't sure where to put the entry point code. We don't want to create duplicated code across the accounts, contacts, etc. The entrypoint code is often put in the root folder of a module eg "removeme" => "modules/Campaigns/RemoveMe.php".

You added the method getConfirmOptInTickFromSugarEmailAddressField was added to display the opt in tick. So in order to create a general way to display the tick that allows child classes to display it, I separated how to check the opt in status and the displayOptInFromSugarEmailAddressField method to handle the display logic.  Perhaps this should be moved. 

So the reason I thought it should be in the Basic.php is because it is related to the code we added in the Basic.php. It is the general case for displaying the tick in an email address field. I assumed, when you create the method in the basic case, you where attempting to make possible for all beans to have an email address field.

Reply:
Ok, but in this case why we needed a new class. Why didn't you just use the exists `Basic`, using new method(s)?

Reply:
anyway, I outsourced the switch..case to a separated method so if you want to override only the requested methods you can extends the class.